lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -81,16 +81,18 @@ def fragment_cache(cached_hash, non_cached_hash)
 
         def serializable_hash_for_collection(options)
           hash = { data: [] }
+          included = []
           serializer.each do |s|
             result = self.class.new(s, instance_options.merge(fieldset: fieldset)).serializable_hash(options)
             hash[:data] << result[:data]
+            next unless result[:included]
 
-            if result[:included]
-              hash[:included] ||= []
-              hash[:included] |= result[:included]
-            end
+            included |= result[:included]
           end
 
+          included.delete_if { |resource| hash[:data].include?(resource) }
```

Comment:
I believe it is the best we can achieve without refactoring the adapter so that it does not call itself recursively for collections (which I'm in favor of, in a subsequent PR). The scope of this PR was to fix a behavior breaking JSON API spec, without modifying the adapter too much.